src/protocols/echo/EchoClient.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements an object for a CHIP Echo unsolicitied
+ *      initiator (client).
+ *
+ */
+
+#include "Echo.h"
+
+namespace chip {
+namespace Protocols {
+
+CHIP_ERROR EchoClient::Init(ExchangeManager * exchangeMgr)
+{
+    // Error if already initialized.
+    if (ExchangeMgr != nullptr)
+        return CHIP_ERROR_INCORRECT_STATE;
+
+    ExchangeMgr            = exchangeMgr;
+    OnEchoResponseReceived = nullptr;
+    ExchangeCtx            = nullptr;
+
+    return CHIP_NO_ERROR;
+}
+
+void EchoClient::Shutdown()
+{
+    if (ExchangeCtx != nullptr)
+    {
+        ExchangeCtx->Abort();
+        ExchangeCtx = nullptr;
+    }
+}
+
+CHIP_ERROR EchoClient::SendEchoRequest(NodeId nodeId, System::PacketBuffer * payload)
+{
+    // Discard any existing exchange context. Effectively we can only have one Echo exchange with
+    // a single node at any one time.
+    if (ExchangeCtx != nullptr)
+    {
+        ExchangeCtx->Abort();
+        ExchangeCtx = nullptr;
+    }
+
+    // Create a new exchange context.
+    ExchangeCtx = ExchangeMgr->NewContext(nodeId, this);
+    if (ExchangeCtx == nullptr)
+    {
+        System::PacketBuffer::Free(payload);
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    return SendEchoRequest(payload);
+}
+
+CHIP_ERROR EchoClient::SendEchoRequest(System::PacketBuffer * payload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Send an Echo Request message.  Discard the exchange context if the send fails.
+    err = ExchangeCtx->SendMessage(kChipProtocol_Echo, kEchoMessageType_EchoRequest, payload);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ExchangeCtx->Abort();
+        ExchangeCtx = nullptr;
+    }
+
+    return err;
+}
+
+void EchoClient::OnMessageReceived(ExchangeContext * ec, const PacketHeader & packetHeader, uint32_t protocolId, uint8_t msgType,
+                                   System::PacketBuffer * payload)
+{
+    EchoClient * echoApp = static_cast<EchoClient *>(ec->GetAppState());
```

Comment:
Do not need to use `AppState`, `this` is the `EchoClient`. Same for `EchoServer` if we can register the delegate instead of callback function ptr and appState.

Reply:
delegate pattern mandate the the client to implement the callback. With this design, the client or server can choose to if  receive the response.